src/mlpack/tests/main_tests/preprocess_scale_test.cpp

```diff
@@ -0,0 +1,117 @@
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#define BINDING_TYPE BINDING_TYPE_TEST
+
+#include <mlpack/core.hpp>
+static const std::string testName = "PreprocessScale";
+
+#include <mlpack/core/util/mlpack_main.hpp>
+#include <mlpack/methods/preprocess/preprocess_scale_main.cpp>
+
+#include "test_helper.hpp"
+#include <boost/test/unit_test.hpp>
+#include "../test_tools.hpp"
+
+#include <cmath>
+
+using namespace mlpack;
+
+struct PreprocessScaleTestFixture
+{
+ public:
+  PreprocessScaleTestFixture()
+  {
+    // Cache in the options for this program.
+    CLI::RestoreSettings(testName);
+  }
+
+  ~PreprocessScaleTestFixture()
+  {
+    // Clear the settings.
+    bindings::tests::CleanMemory();
+    CLI::ClearSettings();
+  }
+};
+
+BOOST_FIXTURE_TEST_SUITE(PreprocessScaleMainTest,
+                         PreprocessScaleTestFixture);
+
+arma::mat dataset = "-1 -0.5 0 1;"
+                    "2 6 10 18;";
+/**
+ * Check that two different scalers give two different output.
+ */
+BOOST_AUTO_TEST_CASE(TwoScalerTest)
+{
+  std::string method = "max_abs_scaler";
+  // Input custom data points.
+  SetInputParam("input", dataset);
+  SetInputParam("scaler_method", std::move(method));
+
+  mlpackMain();
+  arma::mat max_abs_scaler_output = CLI::GetParam<arma::mat>("output");
+
+  method = "standard_scaler";
+  SetInputParam("input", dataset);
+  SetInputParam("scaler_method", std::move(method));
+
+  mlpackMain();
+  arma::mat standard_scaler_output = CLI::GetParam<arma::mat>("output");
+
+  CheckMatricesNotEqual(standard_scaler_output, max_abs_scaler_output);
+}
+
+/**
+ * Check that two different option for a particular scaler give two
+ * different output.
+ */
+BOOST_AUTO_TEST_CASE(TwoOptionTest)
+{
+  std::string method = "min_max_scaler";
+  // Input custom data points.
+  SetInputParam("input", dataset);
+  SetInputParam("scaler_method", method);
+
+  mlpackMain();
+  arma::mat output = CLI::GetParam<arma::mat>("output");
+
+  SetInputParam("input", dataset);
+  SetInputParam("scaler_method", std::move(method));
+  SetInputParam("min_value", 2);
+  SetInputParam("max_value", 4);
+
+  mlpackMain();
+  arma::mat output_with_param = CLI::GetParam<arma::mat>("output");
+
+  CheckMatricesNotEqual(output, output_with_param);
+}
+
+/**
+ * Check that passing unrelated option don't change anything.
+ */
+BOOST_AUTO_TEST_CASE(UnrelatedOptionTest)
+{
+  std::string method = "standard_scaler";
+  // Input custom data points.
+  SetInputParam("input", dataset);
+  SetInputParam("scaler_method", std::move(method));
+  SetInputParam("min_value", 2);
+  SetInputParam("max_value", 4);
+  SetInputParam("epsilon", 0.005);
+  mlpackMain();
+  arma::mat output = CLI::GetParam<arma::mat>("output");
+  arma::mat scaled = "-1.18321596 -0.50709255  0.16903085 1.52127766;"
+                   "-1.18321596 -0.50709255  0.16903085 1.52127766;";
+  CheckMatrices(scaled, output);
+}
+
```

Comment:
I can think of just a couple more tests to add here---

 1. We should check to make sure a trained scaler model gives the same output in two separate runs.  So, for instance, train a scaler model on a dataset and then save `output` and `output_model`, then run again with the same dataset and `input_model` set to the output model of the last run, and make sure the output of the second run is the same as the output of the first.
 2. Make sure `min_value` and `max_value` make a difference when `min_max_scaler` is used.
 3. Try running in regular mode and inverse mode and make sure that either different results are given, or that we recover the original matrix.
 4. Make sure `epsilon` makes a difference for PCA whitening.
 5. Make sure all six strategies give some kind of output and don't crash.  (Maybe this can be combined with (3)).
 6. Make sure invalid values of `epsilon` cause an exception.
 7. Make sure invalid values of `scaler_method` cause an exception.
 8. Make sure an exception is thrown when `max_value` < `min_value` (or they are equal). :)

It can be kind of tedious to write these tests but if you don't then our users will find when something is wrong and write us bug reports. :)  Better to know that we won't get those now.

Reply:
> I can think of just a couple more tests to add here---
> 
>     1. We should check to make sure a trained scaler model gives the same output in two separate runs.  So, for instance, train a scaler model on a dataset and then save `output` and `output_model`, then run again with the same dataset and `input_model` set to the output model of the last run, and make sure the output of the second run is the same as the output of the first.
> 
Added this test.
>     2. Make sure `min_value` and `max_value` make a difference when `min_max_scaler` is used.
> 
It was aldeardy there, check test no 2 :)
>     3. Try running in regular mode and inverse mode and make sure that either different results are given, or that we recover the original matrix.
> 
Added this too, Thanks for the suggestion.
>     4. Make sure `epsilon` makes a difference for PCA whitening.
> 
Wrote a test for this too.
>     5. Make sure all six strategies give some kind of output and don't crash.  (Maybe this can be combined with (3)).
> 
I don't think so , we should write one for this because , I have already checked Transform and InverseTransfrom for the functionality, If you still want me to write , I will do so, Do i have to write for all 6 scalers ?
>     6. Make sure invalid values of `epsilon` cause an exception.
> 
Yes, Added test for Invalid epsilon.
>     7. Make sure invalid values of `scaler_method` cause an exception.
> 
Added test for this too.
>     8. Make sure an exception is thrown when `max_value` < `min_value` (or they are equal). :)
> 
Thanks for the idea, Wrote a test accordingly.
> It can be kind of tedious to write these tests but if you don't then our users will find when something is wrong and write us bug reports. :) Better to know that we won't get those now.

I agree with you, No issue with writing test for any of these. If you have some more let me know, I would be happy enough to implement those.

Reply:
Awesome, thanks for adding these.

> I don't think so , we should write one for this because , I have already checked Transform and InverseTransfrom for the functionality, If you still want me to write , I will do so, Do i have to write for all 6 scalers ?

The thought here is just to check that the user can specify each of the six options for the strategies without there being some crash.  Sometimes it can happen that the documentation says a certain option should be used but then it is spelled differently in the code. :)  So the check is not that the strategy works; in fact, just running `BOOST_REQUIRE_NO_THROW(mlpackMain())` should be fine for each.

Reply:
> It was aldeardy there, check test no 2 :)

You're right, I see it now. :)  Must have missed it the first time.